Create freertos port

leanfrancucci opened the pull request.

leanfrancucci committed:
Added freertos as submodule

cmancon committed:
Added initial port files

Added the FreeRTOS port files created by DaBa and modified rkhplat and rkhtype header files to impact the addition.

cmancon committed:
Corrected erroneous relative location for platform specific files

cmancon committed:
Created a shell of a project to validate the port.

cmancon committed:
Added bsp related files to demo project

cmancon committed:
Included bsp specific files and functions.

cmancon committed:
Removed some unneeded hardware initialisation

cmancon committed:
Included optional info to improve FreeRTOS debug experience

cmancon committed:
Redefinition of startupTask to remove double trace openning

cmancon committed:
Manipulations to get an stable and working demo on CIAA-NXP

dariosb commented on demo/libbsp/platform/arm-cortex/arm_cm4f/ciaa-nxp/bsp_shared.c:
```diff
@@ -383,5 +383,16 @@ rkh_hook_exit( void )
 	RKH_TRC_FLUSH();
 }
 
+void
```
Perfecto, esa era la idea de redefinirla acá. @cmancon pudiste validar si funciona?

cmancon replied:
Con ese cambio (la redefinición) no lo dejé corriendo mucho más que unos minutos pero sí con una modificación temporal en la versión _weak_. Ahora lo puedo dejar un par de horas ejecutando estrictamente lo del commit.

cmancon replied:
Se comprobó que funcionaba correctamente. Cierro la conversación.

leanfrancucci commented:
Después de hacer un análisis de shared y del port para freertos, aporto algunas ideas para resolverlo:
1. Sabemos que `main()` no puede cambiarse porque es una app cross, tenemos que lidiar desde el port (rkhport.c) y el bsp (bsp_shared.c)
2. En bsp_shared.c podemos plantear `bsp_init()` como:
```c
bsp_init(...)
{
    ...
    rkh_fwk_init();
    RKH_FILTER_*;
    RKH_TRC_OPEN();
}
```
3. En rkhport.c:
- `rkh_fwk_init()` es responsable de establecer dónde se ejecuta el tick de RKH (`RKH_TIM_TICK()`) para manejar los timers. En freertos tenemos tres posibilidades:
    - Desde la callback de un software timer de freertos. Esta se ejecuta en contexto de tarea. _Es mi opción favorita_
    - Desde una tarea específica que ejecuta un loop con un retardo. Similar a lo que hacía `rkh_fwk_enter()`
    - Desde el hook de la función tick de freertos. Esta se ejecuta en contexto de ISR
```c
rkh_fwk_init(...)
{
    /* There are 3 alternatives to execute RKH_TIM_TICK() */
    /* 1. From a FreeRTOS software timer (Task context) */ /* <- favorite */
    /* 2. From a specific task (Task context) */
    /* 3. From FreeRTOS tick hook function (ISR context) */
}
```
- rkh_fwk_enter() es responsable principalmente de llamar al scheduler del OS
```c
rkh_fwk_enter(...)
{
    RKH_HOOK_START();   /* start-up callback */
    RKH_TR_FWK_EN();
    vTaskStartScheduler();
    RKH_HOOK_EXIT();    /* cleanup callback */
}
```

cmancon commented:
> Después de hacer un análisis de shared y del port para freertos, aporto algunas ideas para resolverlo:
> 
> 1. Sabemos que `main()` no puede cambiarse porque es una app cross, tenemos que lidiar desde el port (rkhport.c) y el bsp (bsp_shared.c)
> 2. En bsp_shared.c podemos plantear `bsp_init()` como:
> 
> ```c
> bsp_init(...)
> {
>     ...
>     rkh_fwk_init();
>     RKH_FILTER_*;
>     RKH_TRC_OPEN();
> }
> ```
> 
> 1. En rkhport.c:
> 
> * `rkh_fwk_init()` es responsable de establecer dónde se ejecuta el tick de RKH (`RKH_TIM_TICK()`) para manejar los timers. En freertos tenemos tres posibilidades:
>   
>   * Desde la callback de un software timer de freertos. Esta se ejecuta en contexto de tarea. _Es mi opción favorita_
>   * Desde una tarea específica que ejecuta un loop con un retardo. Similar a lo que hacía `rkh_fwk_enter()`
>   * Desde el hook de la función tick de freertos. Esta se ejecuta en contexto de ISR
> 
> ```c
> rkh_fwk_init(...)
> {
>     /* There are 3 alternatives to execute RKH_TIM_TICK() */
>     /* 1. From a FreeRTOS software timer (Task context) */ /* <- favorite */
>     /* 2. From a specific task (Task context) */
>     /* 3. From FreeRTOS tick hook function (ISR context) */
> }
> ```
> 
> * rkh_fwk_enter() es responsable principalmente de llamar al scheduler del OS
> 
> ```c
> rkh_fwk_enter(...)
> {
>     RKH_HOOK_START();   /* start-up callback */
>     RKH_TR_FWK_EN();
>     vTaskStartScheduler();
>     RKH_HOOK_EXIT();    /* cleanup callback */
> }
> ```

cmancon commented:
Perdón, botón equivocado... :S

leanfrancucci commented:
@cmancon respecto del main() y su relación con el port. 
Si la aplicación está basada en el framework entonces el main() generalmente tiene la forma del de shared. Y además se relaciona de manera similar con el bsp y el port.

Ahora, a esta arquitectura se le pueden sumar, sin ningún problema, tareas que no usan el framework.

Por otro lado, si se requiere usar el framework con una aplicación *existente* que usa freertos entonces puede que el port deba modificarse o incluso se requiera crear otro diferente. Eso lo veremos cuando nos enfrentemos con el problema. Por ahora me inclino por resolver el port para aplicaciones basadas en el fwk. 

Respecto de usar el fwk en aplicaciones ya funcionando, es exactamente el caso de lo que podríamos enfrentar al incluir RKH en [ESP8266_RTOS_SDK](https://docs.espressif.com/projects/esp8266-rtos-sdk/en/latest/index.html). Dicho sea de paso, estuve analizando cómo crear tareas propias dentro de la arquitectura ESP8266_RTOS_SDK y en principio parece viable agregar el fwk para resolver la "lógica de más alto nivel" del dispositivo, haciendo uso de todas las bibliotecas y funcionalidades que provee ESP8266_RTOS_SDK

> Por otra parte, tu análisis sugiere que no es necesaria la función rkh_startupTask(). ¿Entendí bien?
Si, en principio no es necesaria.

cmancon committed:
Modified fwk initialisation and FreeRTOS' Scheduler starting point

> rkh_fwk_enter() now starts the FreeRTOS' Scheluder.
> The RKH Timer Tick is generated through a FreeRTOS' Software Timer.
> Removed rkh_startupTask().

leanfrancucci commented on source/portable/freertos/v10.3.0/rkhport.c:
```diff
@@ -129,36 +136,39 @@ rkh_startupTask(void *pvParameter)
+			&( xTimerBuffers[0] )
+	);
+
+    RKH_ASSERT(tick_TimerHandle);
```
Para que sea más entendible el código podes usar RKH_ENSURE() que verifica las post-condiciones

cmancon replied:
Perfecto! Lo incluyo en el próximo commit.

leanfrancucci commented on source/portable/freertos/v10.3.0/rkhport.c:
```diff
@@ -129,36 +136,39 @@ rkh_startupTask(void *pvParameter)
-    TaskHandle_t TaskHandle = NULL;
+	RKH_TRC_OPEN();
+
+#if defined(RKH_USE_TRC_SENDER)
```
:+1:

cmancon replied:
¿Es correcto dejarlo con ```#if defined(RKH_USE_TRC_SENDER)``` o se podría mejorar?

dariosb replied:
Fijate que está relacionado con 
https://github.com/vortexmakes/RKH/blob/45c1128fb646933448a14367a02c6e8bf0b4724b/source/portable/freertos/v10.3.0/rkhport.c#L68
Si no está definido RKH_USE_TRC_SENDER no tiene utilidad publicar el simbolo y tampoco su existencia.

cmancon replied:
Sí, pero es **tal vez** más robusto envolver con ```#if RKH_CFG_TRC_EN == RKH_ENABLED```. ¿Valdrá la pena?

dariosb replied:
No hace falta. Para que RKH_USE_TRC_SENDER sea defined, TRC tiene que ser Enabled.
https://github.com/vortexmakes/RKH/blob/45c1128fb646933448a14367a02c6e8bf0b4724b/source/fwk/inc/rkhitl.h#L74

cmancon commented on demo/cross/shared/main.c:
```diff
@@ -67,9 +67,12 @@ main(int argc, char *argv[])
                          CLI_STK_SIZE);
     }
-
+#ifdef __FREERTOS_V10_03_00__
```
Resuelto en https://github.com/vortexmakes/RKH/pull/13/commits/45c1128fb646933448a14367a02c6e8bf0b4724b

leanfrancucci commented:
@cmancon tenés algo de estadística de `shared` ejecutándose al cabo de un tiempo pronunciado, tipo valores máximos o promedio de `Stack Usage` y `Runtime`?

cmancon commented:
No recuerdo si tengo guardados los resultados pero he dejado funcionando el ejemplo unas 5 horas verificando los valores en las herramientas de Debug de FreeRTOS de MCUXpresso. Me parecían bastante saludables.

Lo que no recuerdo que tuviera era el _watermark_, pero si el uso de `Stack `por tarea y su respectivo `Runtime`

¿Te parece que hagamos un ensayo y registremos acá los resultados?

cmancon committed:
Minor changes as discussed in PR#13

cmancon committed:
Improved BSP layer

> Added "console output" via Semihosting to print info about the Demo.
> Added control of LEDs to show activity.
> Removed unused or unrelated code.

cmancon committed:
Fixed some absolute paths in MCUXpresso project

cmancon committed:
Added OpenOCD config file

cmancon committed:
Improved Debug configurations

> Modified OpenOCD config to manipulate configuration from MCUXpresso.
> Added FreeRTOS helpers to improve integration with the built-in debugger and with OpenOCD.
> Separated into two Build configurations:
  - Debug: no semihosting.
  - Debug_with_Semihosting: Using Redlib(semihost).
> Created four Debug launchers:
  - Debug: No semihosting, No OpenOCD-FreeRTOS integration.
  - Debug Semihosting: Uses Debug_with_Semihosting build but No OpenOCD-FreeRTOS integration.
  - Debug OpenOCD+FreeRTOS: Uses Debug build but No OpenOCD-FreeRTOS integration.
  - Debug OpenOCD+FreeRTOS+Semihosting: Uses Debug_with_Semihosting build and OpenOCD-FreeRTOS integration.

  The OpenOCD-FreeRTOS integration enables tasks listed as threads by GDB. This allows to have fast and easy access to backtrace and current status of each task.

  The built-in debugger benefits with the helpers as it can access more data from the RTOS and show it on dedicated views. For example, depending on the memory scheme being used, it can show heap usage; or make a listing of timers, etc.

leanfrancucci commented:
@cmancon tenés info de ocupación de RAM y ROM de los archivos de la app shared y del fwk?. Tal vez tengas el IDE resume estos datos o sino pueden estar en el archivo que especifica el mapa de carga

cmancon commented:
Bueno, habría que ver qué compilación tomar como parámetro.

Como para dar una idea, el siguiente es el uso de memoria del Shared usando semishoting con `redlib(semihost)` y teniendo el trace activado:
| Memory region | Used Size | Region Size | %age Used |
|--------------:|----------:|------------:|----------:|
|   MFlashA512: |   43504  B |      512 KB |     8.30% |
|   MFlashB512: |       0  B |      512 KB |     0.00% |
|     RamLoc32: |   10200  B |       32 KB |    31.13% |
|     RamLoc40: |     40 KB |       40 KB |   100.00% |
|     RamAHB32: |       0  B |       32 KB |     0.00% |
|     RamAHB16: |      8 KB |       16 KB |    50.00% |
| RamAHB_ETB16: |       0  B |       16 KB |     0.00% |

Tanto RamLoc40 como RamAHB16 estan manejados por el MCUXpresso con el `Managed Linker Script` y _allocan_ `heap `y `stack `respectivamente.

Como estamos usando el [heap3](https://www.freertos.org/a00111.html#heap_3) que envuelve el `malloc()` y el `free()` no puedo aproximar cuánto del heap dedicado es utilizado.

cmancon committed:
Untracked 'settings' folder from mcuxpresso project folder

cmancon committed:
Tidied FreeRTOSConfig file and included helper function to check heap

cmancon committed:
Added Pause function using TECs or DINs depending on the board used

> Included a minimal debounce implementation to read imputs from GPIO.
> Include Pause function trigered by TEC1 in EDU-CIAA or DIN0 in CIAA.

cmancon committed:
Modified conditional compilation with environment variable

> Included the "BOARD" Environment Variable. It is also used to conditionally compile the code to use the specific platform. Its values are compatible with sAPI definitions.
   Further explanation: In case that you use a EDU-CIAA-NXP, the OpenOCD config file depends on a environment variable to know it. If it is not defined, it will fallback to CIAA-NXP. To give a seamless experience, this variable is included in the project and, additionally, it's used to adapt the GPIO pins hardware configuration.

> All Build Configurations were checked and the "Release" configuration was corrected to work properly.

cmancon committed:
Updated Debug Configurations to make explicit use of both platforms

> Created debug configurations (and launchers) for each platform.
> Simplified options to use OpenOCD FreeRTOS integration and semihosting in any case.
> Left a "legacy" configuration as a temporary backup.

dariosb committed:
fix freertos submodule link

dariosb committed:
Adds BOARD definition to project file

dariosb committed:
Solve environment variable definition

cmancon committed:
Deleted FreeRTOS (general repository) submodule

cmancon committed:
Included FreeRTOS kernel-only repository as a submodule

> Head pointing to tag V10.3.0-kernel-only.

leanfrancucci commented:
Are there pending tasks related to this port?. It would be great to turn over a new leaf, and then to think about fresh application examples with freeRTOS and RKH, as we have talked about several weeks ago :muscle: :muscle:

cmancon commented:
We are back to English? hehe
Possible tasks to finish:
1. Merge #14  
2. Compose some sort of memory usage report.
3. Compose a README.

What do you think?

cmancon committed:
Moved bsp_timeTick() implementation out of conditional compilation block

> Before implementing the tact switch (or digital inputs depending the specific platform) bsp_timeTick() was only used by the trace output. For that reason, the implementation was inside a conditional compilation block.
   When the previously said functionality was implemented, the corresponding section of the bsp_timeTick() function was modified to be conditionally compiled but  it wasn't noticed the the entire function was inside a conditional block itself.
   This situation is corrected here and will correctly compile if the trace functionality is disabled.

cmancon committed:
Merge pull request #14 from vortexmakes/issue-freertos-submodule

fix freertos submodule link

cmancon committed:
Update .travis.yml

cmancon committed:
Update .travis.yml

cmancon committed:
Update .travis.yml

cmancon committed:
First iteration on the Shared demo Readme
